manager/state/raft/storage/snapwrap.go

```diff
@@ -0,0 +1,196 @@
+package storage
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/coreos/etcd/pkg/fileutil"
+	"github.com/coreos/etcd/raft/raftpb"
+	"github.com/coreos/etcd/snap"
+	"github.com/pkg/errors"
+)
+
+// This package wraps the github.com/coreos/etcd/snap package, and encodes
+// the bytes of whatever snapshot is passed to it, and decodes the bytes of
+// whatever snapshot it reads.
+
+// Snapshotter is the interface presented by github.com/coreos/etcd/snap.Snapshotter that we depend upon
+type Snapshotter interface {
+	SaveSnap(snapshot raftpb.Snapshot) error
+	Load() (*raftpb.Snapshot, error)
+}
+
+// SnapFactory provides an interface for the different ways to get a Snapshotter object.
+// For instance, the etcd/snap package itself provides this
+type SnapFactory interface {
+	New(dirpath string) Snapshotter
+}
+
+var _ Snapshotter = &wrappedSnap{}
+var _ Snapshotter = &snap.Snapshotter{}
+var _ SnapFactory = snapCryptor{}
+
+// wrappedSnap wraps a github.com/coreos/etcd/snap.Snapshotter, and handles
+// encoding/decoding.
+type wrappedSnap struct {
+	*snap.Snapshotter
+	encoder  Encoder
+	decoders map[Algorithm]Decoder
+	mu       sync.RWMutex
+}
+
+// SaveSnap encodes the snapshot data (if an encoder is exists) before passing it onto the
+// wrapped snap.Snapshotter's SaveSnap function.
+func (s *wrappedSnap) SaveSnap(snapshot raftpb.Snapshot) error {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	toWrite := snapshot
```

Comment:
We can't, really.  :|  Since we're doing the migrations to a new directory, I've updated the migration functions to also just always wrap the data in an `WrappedRecord`.  So we're going to assume any snapshots in the new directory can always be deserialized into such a structure.